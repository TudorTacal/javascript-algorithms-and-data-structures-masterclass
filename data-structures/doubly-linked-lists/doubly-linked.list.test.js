const { Node, Dll } = require('./doubly-linked-list');

describe('dll', () => {
  test('#push should add a node at the end of the dll when 0 nodes', () => {
    let dll = new Dll();
    let expectedResult = {
      head: {
        val: 'Hello',
        next: null,
        prev: null,
      },
      tail: {
        val: 'Hello',
        next: null,
        prev: null,
      },
      length: 1,
    };

    dll.push('Hello');

    expect(dll.length).toEqual(1);
    expect(dll).toEqual(expectedResult);
    expect(dll).toMatchSnapshot();
  });

  test('#push should add a node at the end of the dll when multiple nodes', () => {
    let dll = new Dll();

    dll.push('Hello');
    dll.push('World');

    expect(dll.length).toEqual(2);
    expect(dll).toMatchSnapshot();
  });

  test('#pop should remove the last node when multiple nodes', () => {
    let dll = new Dll();

    dll.push('Hello');
    dll.push('World');
    let result = dll.pop();

    expect(dll.length).toEqual(1);
    expect(result).toMatchSnapshot();
    expect(dll).toMatchSnapshot();
  });

  test('#pop should remove the last node when 1 node', () => {
    let dll = new Dll();

    dll.push('Hello');
    let result = dll.pop();

    expect(dll.length).toEqual(0);
    expect(result).toMatchSnapshot();
    expect(dll).toMatchSnapshot();
  });

  test('#shift should remove the first node when 1 node', () => {
    let dll = new Dll();

    dll.push('Hello');
    let result = dll.shift();

    expect(dll.length).toEqual(0);
    expect(result).toMatchSnapshot();
    expect(dll).toMatchSnapshot();
  });

  test('#shift should remove the first node when multiple nodes', () => {
    let dll = new Dll();

    dll.push('Hello');
    dll.push('World');
    let result = dll.shift();

    expect(dll.length).toEqual(1);
    expect(result).toMatchSnapshot();
    expect(dll).toMatchSnapshot();
  });

  test('#unshift should remove the first node when multiple nodes', () => {
    let dll = new Dll();

    dll.push('Hello');
    dll.push('World');
    dll.unshift('Salut');

    expect(dll.length).toEqual(3);
    expect(dll).toMatchSnapshot();
  });

  test('#get should return the node at the specified index', () => {
    let dll = new Dll();

    dll.push('Hello');
    dll.push('Beautiful');
    dll.push('Crazy');
    dll.push('World');

    let result = dll.get(1);
    expect(result).toMatchSnapshot();
  });
});