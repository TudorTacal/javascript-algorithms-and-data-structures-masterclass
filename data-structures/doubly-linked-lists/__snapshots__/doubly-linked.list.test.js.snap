// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`dll #pop should remove the last node when 1 node 1`] = `
Node {
  "next": null,
  "prev": null,
  "val": "Hello",
}
`;

exports[`dll #pop should remove the last node when 1 node 2`] = `
DoublyLinkedList {
  "head": null,
  "length": 0,
  "tail": null,
}
`;

exports[`dll #pop should remove the last node when multiple nodes 1`] = `
Node {
  "next": null,
  "prev": Node {
    "next": null,
    "prev": null,
    "val": "Hello",
  },
  "val": "World",
}
`;

exports[`dll #pop should remove the last node when multiple nodes 2`] = `
DoublyLinkedList {
  "head": Node {
    "next": null,
    "prev": null,
    "val": "Hello",
  },
  "length": 1,
  "tail": Node {
    "next": null,
    "prev": null,
    "val": "Hello",
  },
}
`;

exports[`dll #push should add a node at the end of the dll when 0 nodes 1`] = `
DoublyLinkedList {
  "head": Node {
    "next": null,
    "prev": null,
    "val": "Hello",
  },
  "length": 1,
  "tail": Node {
    "next": null,
    "prev": null,
    "val": "Hello",
  },
}
`;

exports[`dll #push should add a node at the end of the dll when multiple nodes 1`] = `
DoublyLinkedList {
  "head": Node {
    "next": Node {
      "next": null,
      "prev": [Circular],
      "val": "World",
    },
    "prev": null,
    "val": "Hello",
  },
  "length": 2,
  "tail": Node {
    "next": null,
    "prev": Node {
      "next": [Circular],
      "prev": null,
      "val": "Hello",
    },
    "val": "World",
  },
}
`;

exports[`dll #shift should remove the first node when 1 node 1`] = `
Node {
  "next": null,
  "prev": null,
  "val": "Hello",
}
`;

exports[`dll #shift should remove the first node when 1 node 2`] = `
DoublyLinkedList {
  "head": null,
  "length": 0,
  "tail": null,
}
`;

exports[`dll #shift should remove the first node when multiple nodes 1`] = `
Node {
  "next": null,
  "prev": null,
  "val": "Hello",
}
`;

exports[`dll #shift should remove the first node when multiple nodes 2`] = `
DoublyLinkedList {
  "head": Node {
    "next": null,
    "prev": null,
    "val": "World",
  },
  "length": 1,
  "tail": Node {
    "next": null,
    "prev": null,
    "val": "World",
  },
}
`;

exports[`dll #unshift should remove the first node when multiple nodes 1`] = `
DoublyLinkedList {
  "head": Node {
    "next": Node {
      "next": Node {
        "next": null,
        "prev": [Circular],
        "val": "World",
      },
      "prev": [Circular],
      "val": "Hello",
    },
    "prev": null,
    "val": "Salut",
  },
  "length": 3,
  "tail": Node {
    "next": null,
    "prev": Node {
      "next": [Circular],
      "prev": Node {
        "next": [Circular],
        "prev": null,
        "val": "Salut",
      },
      "val": "Hello",
    },
    "val": "World",
  },
}
`;
